File: 3. Longest Substring Without Repeating Characters.java
import java.util.*;
class Solution {
    // solve with a sliding window
    public int lengthOfLongestSubstring(String s) {
        // base case
        int len = s.length();
        if (len == 0) {
            return 0;
        }
        
        // create a hashtable to store chars in current substring
        Hashtable<Character, Integer> t = new Hashtable<>();
        // left pointer
        int j = 0;
        int maxLen = 0;
        // loop over each char with the right pointer
        for (int i = 0; i < len; i++) {
            char c = s.charAt(i);
            // if c not in hash table, put it in
            if (!t.containsKey(c)) {
                t.put(c, 1);
            // if c already in hash table, in order to keep the chars in substring unique
            } else {
                // move the left point to the right
                while (j <= i) {
                    // if j points at a char same as c
                    if (s.charAt(j) == c) {
                        // move j one step more to the left, to exclude the repeating char
                        j++;
                        break;
                    // j hasn't reached the repeating char
                    } else {
                        // remove the char at j from hash table, as it's no longer in the substring
                        t.remove(s.charAt(j));
                        // move one step to the right
                        j++;
                    }
                }
            }
            // substring (j, i) now contains only unique chars. update maxLen if needed
            maxLen = Math.max(maxLen, i - j + 1);
        }
        return maxLen;
    }
}

File: 3. Longest Substring Without Repeating Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        
        # base case:
        if not s:
            return 0
        
        n = len(s)
        max_length = 1
        
        # 2 pointers
        i = 0
        j = 1
        
        while i <= j and j < n:
            
            if s[j] in set(s[i:j]):
                
                while s[i] != s[j]:
                    i += 1
                    # when while loop ends, s[i] == s[j]
                
                # move i one more to the left    
                i += 1
                
            # now substring only contains unique chars
            max_length = max(max_length, len(s[i:j + 1]))
            
            j += 1
                
        return max_length